Migrate from pyjwkest+cryptodome to pyjwt+cryptography

stefanfoulis opened the pull request:
# Use PyJWT+cryptography instead of jwkest+Cryptodrome

- Switches `id_token` signing and verification to use `PyJWT` with `cryptography`
- Removes `jwkest` and `Cryptodome` dependencies
- The `future` dependency, previously required by `jwkest`, is no longer needed. `future` has an unfixed security vulnerability and seems unmaintained. (should fix #444, #293 and #207 )
- adds caching of RSA keys to avoid repeated expensive key loading operations (should fix #374)

Replaces #207
Resolves #444 #293 #207

# Notes

Key caching could be improved if we'd store more stuff on the `RSAKey` model (`kid`, `public_key`) . But I don't want to introduce any database changes at this time.

stefanfoulis committed:
Use PyJWT+cryptography instead of jwkest+Cryptodrome

- Switches id_token signing and verification to use PyJWT with cryptography
- Removes jwkest and Cryptodome dependencies
- `future` dependency previously required by jwkest is no longer needed (it had unfixed security vulnerabilities and seems unmaintained)
- adds caching of RSA keys to avoid repeated expensive key loading operations

juanifioren commented on oidc_provider/tests/cases/test_token_endpoint.py:
```diff
@@ -620,17 +614,15 @@ def test_idtoken_sign_validation(self):
 
-        JWS().verify_compact(response_dic["id_token"].encode("utf-8"), RSAKEYS)
+        # This will raise an exception if verification fails
+        decode_id_token(response_dic["id_token"], self.client)
```
Can we have a test with the exception?

stefanfoulis replied:
looks like we only use `decode_id_token()` in tests and nowhere else.

:thinking: maybe we could remove it entirely?

stefanfoulis replied:
(In the meantime I added a test to check for the exception)

juanifioren replied:
I would keep `decode_id_token`, it was added as a helper util that could be used for specific cases or custom hooks.

juanifioren commented:
We should also investigate the impact on this issue: https://github.com/juanifioren/django-oidc-provider/issues/374

stefanfoulis commented:
> We should also investigate the impact on this issue: #374

Thanks for the hint. A quick test on the new implementation shows that it is still doing a significant amount of CPU intensive stuff when "loading" the keys, like with the old implementation. I'll check to see how we can improve that.

stefanfoulis commented:
@juanifioren I've cleaned it up a bit, added more tests and have addressed the performance issue. I have not yet done any extensive tests in our real-world project yet, but based on unit-tests it should work.

juanifioren commented:
I think we should create a simple test showing that the old library generated these keys with these parameters, and that the new one produces the same output.

I know we no longer have the old library, but we can hardcode the output and it will still serve as a valid test. What do you think?" @stefanfoulis

stefanfoulis commented:
@juanifioren I added reference data from the old version and some test to check they match. Is that roughly what you had in mind?

stefanfoulis commented:
(Let me know before you merge, so I can squash the commits into one)

stefanfoulis commented:
This PR (together with some other changes in our fork) has gone through our QA and has been in use for a few days now... so far no issues.

stefanfoulis commented:
I rebased and squashed the changes in to 1 commit.

juanifioren commented:
@stefanfoulis I have that option when merging, so there’s no need for it. 
Second, after testing the command, I think we should remove `oidc_provider/tests/tools/generate_reference_data.py`, since it depends on a non-existent dependency. I would also remove the compat tests. Sorry for the misunderstanding, I just wanted to double-check that both libraries were generating the same encryption.

stefanfoulis commented:
ok, I'll remove those tests again 👍

stefanfoulis commented:
@juanifioren I removed the compatibility tests.

juanifioren reviewed (approved):
LGTM! great contribution 👏
